### PR TITLE
Pr/selector max specificity

### PR DESCRIPTION
**Pull request by benfrain**
Hi,

So, following on from discussions with @davidtheclark and @jeddy3 I thought it might be more useful to generate a PR at this point. 

This is the first attempt to address: https://github.com/stylelint/stylelint/issues/721

Hopefully that will make it easier to look at this. Forgive me if this is premature - I'm a total newbie to this process!

Any feedback appreciated.

Thanks, Ben

### CODE REVIEW

File: src/rules/selector-max-specificity/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# selector-max-specificity
+
+Control the specificity of selectors.
+
+First define the maximum allowable selector specificity as computed by the [W3C specificity calculator](https://www.w3.org/TR/selectors/#specificity):
```
Comment (davidtheclark) [inline src/rules/selector-max-specificity/README.md]: That link describes the rules of the calculation but isn't itself a "calculator" (is there one? that would be handy ...)

File: src/rules/selector-max-specificity/index.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+export const messages = ruleMessages(ruleName, {
+  expected: (selector, specificity) => `Expected "${selector}" to have a specificity equal to or less than "${specificity}"`,
+})
+
+export default function (max) {
+  return (root, result) => {
+    const validOptions = validateOptions(result, ruleName, {
+      actual: max,
+      possible: [isNumber],
```
Comment (davidtheclark) [inline src/rules/selector-max-specificity/index.js]: Can be more specific with this, right? It has a max of four digits, cannot have decimals, etc.

**Comment by benfrain** [comment]
Would it be worth linking to https://specificity.keegan.st instead or as well as that is a visual version of the module being used to do the calculation?

File: src/rules/selector-max-specificity/index.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    })
+
+    if (!validOptions) { return }
+
+    root.walkRules(checkSpecificity)
+
+    function checkSpecificity(rule) {
+      // using rule.selectors gets us each selector in the eventuality we have a comma separated set
+      rule.selectors.forEach(function (selector) {
+        let computedSpecificity = calculate(selector)[0].specificity.replace(/,/g, "")
```
Comment (davidtheclark) [inline src/rules/selector-max-specificity/index.js]: By convention we've been using `const` on variables that shouldn't change.

Comment (davidtheclark) [inline src/rules/selector-max-specificity/index.js]: Maybe coerce the specificity to a number before the comparison (`const computedSpecifity = Number(..)`), so that if anything goes haywire we might end up with less confusing debugging?

File: src/rules/selector-max-specificity/index.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        // Check if the selector specificity exceeds the allowed maximum
+        if (computedSpecificity > max) {
+          // Use report to output any messages
```
Comment (davidtheclark) [inline src/rules/selector-max-specificity/index.js]: No need for this comment

File: src/rules/selector-max-specificity/index.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+          report({ 
+            ruleName: ruleName,
+            result: result,
+            node: rule,
```
Comment (davidtheclark) [inline src/rules/selector-max-specificity/index.js]: By convention, we've been using ES2015 object shorthands here ... `{ ruleName, result, .. }`

**Comment by davidtheclark** [comment]
Awesome, @benfrain! Good start on one of the most important selector rules.

Yeah, I think it'd be great to add that extra link for explanation.

Looks like it's working as expected so far. Very cool. I'm afraid the road ahead is all about more tests and trying to find edge-cases ... 

Here are some questions and ideas for cases that should be covered:

- If a selector has 11 type selectors (`li > li > li > li >` etc.) does it exceed a `0010` max? Or do you need a class selector?
- Since inline styles won't apply to this, should the examples and explanations all use 3 digit specificities?
- How is it going to work if you have nested rules in your stylesheet?
- What if there's interpolation in the selector? 
- Just to make sure that everything works as expected, you may want to include at least one more test group with a different max value.
- Error with a selector on a line other than 1 to make sure positioning works as expected.
- What will it do with `&` parent selectors?

File: src/rules/selector-max-specificity/README.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
+# selector-max-specificity
```
Comment (jeddy3) [inline src/rules/selector-max-specificity/README.md]: The README for every rule follows [a consistent format](http://stylelint.io/developer-guide/rules/#readme). Have a look at the [README for `max-line-length`](https://github.com/stylelint/stylelint/tree/master/src/rules/max-line-length) as that's very similar to your rule.

File: src/rules/selector-max-specificity/README.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+Control the specificity of selectors.
```
Comment (jeddy3) [inline src/rules/selector-max-specificity/README.md]: "Limit the specificity of selectors." - as an convention we use "Limit" to describe `max-*` rules.

File: src/rules/selector-max-specificity/README.md [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
```
Comment (jeddy3) [inline src/rules/selector-max-specificity/README.md]: Add an example here (between the single line description and the expanded description) to make it clear that this rule applies to to each individual selector in a selector list. Similar to how https://github.com/stylelint/stylelint/tree/master/src/rules/selector-class-pattern#selector-class-pattern does it.

File: src/rules/selector-max-specificity/README.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+First define the maximum allowable selector specificity as computed by the [W3C specificity calculator](https://www.w3.org/TR/selectors/#specificity):
+
+For example:
+
+| inline | ID | Class | type
+|---------|----|---------|-------
+| 0 | 0 | 2 | 0
+
+Set this as a 4 digit value in your Stylelint configuration:
```
Comment (jeddy3) [inline src/rules/selector-max-specificity/README.md]: Follow the standard ["Options"](https://github.com/stylelint/stylelint/tree/master/src/rules/max-line-length#options) layout here i.e. show patterns that are considered warnings and then another set of patterns that are not.

File: src/rules/selector-max-specificity/README.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+Then any selectors that exceed that specificity e.g.
+
+````
+.thing .thing .thing {
+    width: 100%;
+}
+````
+
+Would fail with the message:
```
Comment (jeddy3) [inline src/rules/selector-max-specificity/README.md]: You can remove this.

File: src/rules/selector-max-specificity/README.md [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+````
+"max-selector-specificity": 0020
+````
+
+Then any selectors that exceed that specificity e.g.
+
+````
+.thing .thing .thing {
+    width: 100%;
```
Comment (jeddy3) [inline src/rules/selector-max-specificity/README.md]: No need to provide an declaration in your examples. Each example is kept as terse as possible e.g:
```
.class tag #id {}
```

File: src/rules/selector-max-specificity/index.js (hunk 5 above)
Comment (jeddy3) [inline src/rules/selector-max-specificity/index.js]: FYI @davidtheclark We should be able to lint for this soon once https://github.com/stylelint/stylelint/issues/744 is done :)

**Comment by jeddy3** [comment]
As @davidtheclark said, great start Ben! 

I've nothing more to add to @davidtheclark code comments. So I've added a few comments about the documentation while it's fresh in my mind. Focus on the code stuff first and we can revisit the documentation at the end though.

Some of @davidtheclark's ideas/questions are quite involved and I think others are easier pickings. I suggest starting with the easier pickings first, so that you have a good solid foundation to build on before getting stuck into the more involved ones.  

I reckon the following are the easier pickings:

- Since inline styles won't apply to this, should the examples and explanations all use 3 digit specificities?
- Just to make sure that everything works as expected, you may want to include at least one more test group with a different max value.
- Error with a selector on a line other than 1 to make sure positioning works as expected.
- If a selector has 11 type selectors (`li > li > li > li >` etc.) does it exceed a `0010` max? Or do you need a class selector?

Once you've got those ticked off you can focus on (and we can discuss) these tricker ones:

- How is it going to work if you have nested rules in your stylesheet?
- What will it do with `&` parent selectors?
- What if there's interpolation in the selector? 

If you like, you can chip away at each idea/question in individual commits. It'll help break the remaining tasks into bitesize chunks that will probably be easier to work on (and easier for us to feed back on).

It's looking good so far! It'll be fantastic to see this rule in core :)

File: src/rules/selector-max-specificity/__tests__/index.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    message: messages.expected(".thing div .thing .sausages", 30),
+    line: 1,
+    column: 1,
+  })
+  tr.notOk(".thing div .thing, .sausages .burgers .bacon a {}", {
+    message: messages.expected(".sausages .burgers .bacon a", 30),
+    line: 1,
+    column: 20,
+  })
+})
```
Comment (jeddy3) [inline src/rules/selector-max-specificity/__tests__/index.js]: You might want to [install a `editorconfig` plugin](http://editorconfig.org/#download) for your code editor. It will automatically configure your editor for consistent whitespace e.g. adding an empty line at the bottom of all the files.

**Comment by benfrain** [comment]
Thanks men, will try and look at these this evening/tomorrow. One very important point that @davidtheclark exposes is that eleven elements won't actually make `0011` as one might expect. It would instead be more accurately described as `0,0,0,11` so I need to revisit how the config is set and compared. Would it be OK to set like this instead (rather than `0030`):

````
"selector-max-specificity": "0,0,3,0"
````

That way I can so a like for like comparison with the specificity result.

I suppose we _could_ drop the leading zero (inline styles), just not sure if people are used to seeing that pattern and might be confused by it being different (although in fairness I've also seen it done with four digits and the last one reserved for universal selector that never gets used too).

I'm happy to go with your preference on that?

**Comment by benfrain** [comment]
Sorry, I have some _real_ beginner questions here if you wouldn't mind responding: https://gist.github.com/benfrain/f82590ab6e915d0c4664

**Comment by jeddy3** [comment]
I think we should go with comma separated list of three numbers, as [described in the selector level 4 spec](https://drafts.csswg.org/selectors-4/#specificity-rules) because it's the latest doc I could find on the matter.

```js
"selector-max-specificity": "0,3,0"
```

What do you both think?

As @davidtheclark suggested, we should update the options validator to make validate against this format.

**Comment by benfrain** [comment]
OK, think I've addressed the 'easier pickings' if you'd like to take a look at the latest commit?

**Comment by davidtheclark** [comment]
Looks great, @benfrain.

Here are my suggestions for the other points:

> How is it going to work if you have nested rules in your stylesheet? And what will it do with & parent selectors?

I'm thinking that we should actually be able to construct complete selectors by looking up the hierarchy, right? Recursively look at the a rule's descendants (i.e. at its parents, who looks at its own parent, who looks at its own parent ...), and accumulate all the selectors from any rules found up the ancestral tree. If there is no `&` in the selector, put the ancestral selector in front; if there is a `&`, put it there.

I don't know if this is too naive an approach or not --- but at first thought I think it could work.

> What if there's interpolation in the selector?

This one sucks. I almost feel like we'd just have to ignore any selector with interpolation. How do we know there's interpolation? If it's not a valid CSS identifier, we could assume it's something fancy and just ignore the selector. `$thing` and `#{$thing}` and `$(thing)` are all not valid identifiers. I wrote a regex for this here: https://github.com/stylelint/stylelint/blob/master/src/rules/font-family-name-quotes/index.js#L28 --- it's only drawback is that it doesn't know about escaping. But I think a caveat in the docs could address that.

**Comment by benfrain** [comment]
I'll get the existing logic tidied up inline with the latest improvements and amend the README.md to address earlier comments and then once you're happy at that stage I'll attempt the nesting. Do you know — are there any tests that currently need to look up/down the hierarchy I can look at to get an idea of the basic approach?

**Comment by benfrain** [comment]
Oh, yes, I'll be honest, I have no thoughts of any use on interpolation yet!

**Comment by davidtheclark** [comment]
> Do you know — are there any tests that currently need to look up/down the hierarchy I can look at to get an idea of the basic approach?

There are several rules that look at the immediate parent of a node, e.g. comment-empty-line-before. But the only one I can think of that recursively looks up the whole tree is indentation, and that one's a terrible beast.

I'd say the way to proceed would be to write the tests first, then try to get some logic in place. If you find it too confusing or frustrating, I'd be happy to help out (especially if tests are already in place).

Sound good?

**Comment by benfrain** [comment]
Yes, peachy, thanks :+1:

**Comment by jeddy3** [comment]
> If there is no & in the selector, put the ancestral selector in front; if there is a &, put it there.

We might want to carefully consider what types, if any, of nesting we're going to support. For example, if we're just going to support [direct nesting](http://tabatkins.github.io/specs/css-nesting/#direct) then we don't need to consider putting the ancestral selector in front when there is no `&` selector (SCSS users can use the rule, but only if they always put a `&` in their nested selectors. 

What about [at-nest](http://tabatkins.github.io/specs/css-nesting/#at-nest) nesting? Should we be including the `atRule.param`s of those at-rules in the calculations?

Or do we say that this rule, like `stylelint-selector-bem-pattern`, needs to come after any nesting processor?

I don't think there's an easy answer here.

Ref: https://github.com/stylelint/stylelint/issues/489#issuecomment-156758945 & https://github.com/stylelint/stylelint/issues/498

**Comment by benfrain** [comment]
I don't think we should factor in at-rules as my understanding is they effect applicability, not specificity (it's what is inside the at-rule that adds to the specificity, not the at-rule itself)

I certainly want to include the use of `&` in nested selectors I personally use that pattern heavily (I only allow overrides to a key selector to be nested in the syntax I use/police)

Running this after any nesting processor (I presume at that point everything would be 'normal' CSS) negates the issue but it feels a bit like we'd be cheating a little if we did that. I'd certainly love to see feedback of the specificity in my editor and as I write a nested syntax that would be missing.

**Comment by jeddy3** [comment]
> (it's what is inside the at-rule that adds to the specificity, not the at-rule itself)

I'm referring specifically to the [`@nest` at-rule](http://tabatkins.github.io/specs/css-nesting/#at-nest), which is used when the `&` isn't the first thing in a nested selector. e.g.

```css
a {
 
  /* direct nesting */
  & b {}

  /* @nest testing must be used here when using postcss-nesting/postcss-cssnext */
  @nest b & {}
}
```

> I'd certainly love to see feedback of the specificity in my editor

I can see the appeal :)

Perhaps supporting just [*direct nesting*](http://tabatkins.github.io/specs/css-nesting/#direct) is a good first step as it's common to SCSS, `postcss-nested` and `postcss-nesting`. That feels like a sensible compromise and one that can concisely explained in the docs.

**Comment by benfrain** [comment]
> Perhaps supporting just direct nesting is a good first step as it's common to SCSS, postcss-nested and postcss-nesting. That feels like a sensible compromise and one that can concisely explained in the docs.

I think that's probably very practical and I would venture the most applicable to most users (e.g. plenty of people use Sass nesting, I don't personally know anyone using or advocating the use of the CSS draft nesting as it's not settled)

I must admit, figuring out the nesting part feels a bit over my skill level at present. I'll likely get the tests in and everything else I'm confident-ish to sort (tests, README etc) and hand over to someone far more competent *cough* @davidtheclark *cough* *cough* @jeddy3  *cough* :+1:

**Comment by benfrain** [comment]
In my naive head I was thinking of checking if the thing was root and then if not, push each selector into an array I would eventually flatten to a string to test. However, I'm pretty sure there are a bazillion possibilities I haven't considered.

**Comment by davidtheclark** [comment]
I'm hoping there's not :) But if you write the tests, I'll do what I can to make them pass!

**Comment by jeddy3** [comment]
> I don't personally know anyone using or advocating the use of the CSS draft nesting as it's not settled

I do :) Also, [`postcss-nesting'](https://www.npmjs.com/package/postcss-nesting) was downloaded 38000 times this month (probably due to its inclusion in `postcss-cssnext`) and some of those people are probably using `@nest` too.

It's kind of a moot point though as I think we've, correctly, settled on direct nesting as a good start :)

> A style rule can be directly nested within another style rule if its selector is nest-prefixed. To be nest-prefixed, a nesting selector must be the first simple selector in the first compound selector of the selector. If the selector is a list of selectors, every complex selector in the list must be nest-prefixed for the selector as a whole to nest-prefixed.

Which, I think, is a kind of fancy way of saying each nested selector must begin with either `&`, `& ` `&.`, `&#`, `&[` or `&:`, right?

So, I reckon we start with some tests for those and see where we end up :)

**Comment by benfrain** [comment]
OK, just pushed some more revisions. Hopefully the README is more inline with expectations now.

While the regex @davidtheclark suggested looks like it should work (and indeed all the regex testers I tried it with online seemed to work with it) in practice it didn't. Probably something obviously stupid I'm doing but for now I have kept it as it was before.

I've left the start I made on nesting out as in likelihood you'd just need to delete it anyway ;) However, I've added a few nested tests in so I hope this helps a little. Sorry I couldn't sort this bit out!

**Comment by davidtheclark** [comment]
I contributed a commit over here: https://github.com/stylelint/stylelint/pull/761 Do not know a better way in git to handle this collaboration ...

**Comment by jeddy3** [comment]
> Actually the @nest rule should be easy to support with the resolve-nested-selector module.

That's fantastic! I saw the tests when you published yesterday and thought they were super sweet :)

> Regarding interpolation: I think we need to just run away when we see it. 

Agreed.

---

Finally, @benfrain can you please add a simple *direct nesting* and a simple `@nest` pass and fail pattern to the README e.g.

Fail with `1,1,0`:

```css
/* direct */
.class {
  & #id.class2 {}
}
```

```css
/* @nest */
.class {
  @nest #id.class2 & {}
}
```

Even though `postcss-resolve-nested-selector` supports things like:

- [Unwrapping rules without ampersands](https://github.com/davidtheclark/postcss-resolve-nested-selector/blob/master/test/postcss-nested-test.js#L18).
- [Replacing ampersands in strings](https://github.com/davidtheclark/postcss-resolve-nested-selector/blob/master/test/postcss-nested-test.js#L54).
* [Unwrapping nested at rules](https://github.com/davidtheclark/postcss-resolve-nested-selector/blob/master/test/postcss-nested-test.js#L102).

And they can use when using `postcss-nested` and scss, I think we should (as we've done else where) only *document* standard-*ish* nesting features. 

This rule is coming along great, and I think it's going to be good one :)
